### PR TITLE
Cleanup

### PR DESCRIPTION
**Pull request by DavidBurkett**
Sorry this is so large. I got a bit carried away. Fortunately, most of the changes are just moving code around and removing duplicate code.

At a high-level, this PR:
* Changes kernel sorting to include pegin kernels first.
* Starts cleaning up state validation. Another PR will follow.
* Switches to just TxBuilder instead of mixed TxBuilder/tx::Builder for unit-tests.
* Removes OutputFactory & KernelFactory in favor of factory methods directly on the objects themselves.
* Adds a `change_output` attribute to the libmw::Coin which will be useful when determining how to display txs in the UI.

### CODE REVIEW

File: src/wallet/Wallet.cpp [hunk 1]
```diff
@@ -166,72 +162,47 @@ void Wallet::BlockDisconnected(const mw::Block::CPtr& pBlock)
 }
 
 // TODO: Implement
-void Wallet::ScanForOutputs(const libmw::IChain::Ptr&)
+void Wallet::ScanForOutputs(const libmw::IChain::Ptr& pChain)
```
Comment (DavidBurkett) [inline src/wallet/Wallet.cpp:165]: This function is still unfinished. You can ignore it for now.

File: src/wallet/PegOut.cpp [hunk 2]
```diff
@@ -29,11 +26,11 @@ mw::Transaction::CPtr PegOut::CreatePegOutTx(
 
     // Create change output.
     // We randomly generate the sender_key and output_blind.
-    // Receiver key is generate by OutputFactory using the wallet's stealth address.
+    // Receiver key is generate by Output::Create using the wallet's stealth address.
```
Comment (hectorchu) [inline src/wallet/PegOut.cpp:29]: generated by

File: test/framework/include/test_framework/models/TxOutput.h [hunk 3]
```diff
@@ -16,69 +16,21 @@ class TxOutput
...
-
-        return OwnerData(
-            features,
-            std::move(sender_pubkey),
-            std::move(receiver_pubkey),
-            std::move(R),
-            std::move(encrypted_data),
-            std::move(signature)
-        );
+        return TxOutput{ blinding_factor, amount, std::move(output) };
```
Comment (hectorchu) [inline test/framework/include/test_framework/models/TxOutput.h:26]: std::move(output) but parameter is const &.

**Review by hectorchu — APPROVED** [review]
Looks good. I didn't notice any problems apart from a couple of comments.

**Comment by hectorchu** [comment]
Actually I just looked again and I think the change_output stuff is incomplete. For instance it's deserialized from the decrypted data in RewindOutput but it's not serialized in OwnerData::Create.

**Comment by DavidBurkett** [comment]
> Actually I just looked again and I think the change_output stuff is incomplete. For instance it's deserialized from the decrypted data in RewindOutput but it's not serialized in OwnerData::Create.

I've completely replaced the output encoding logic to resemble https://gist.github.com/tevador/f3a66a2f15a8a3a04a1dde1ea65f9205

File: test/framework/include/test_framework/models/TxOutput.h (hunk 3 above)
Comment (DavidBurkett) [reply test/framework/include/test_framework/models/TxOutput.h:26]: Fixed in a new PR I'll be submitting in a few minutes.

File: src/wallet/PegOut.cpp (hunk 2 above)
Comment (DavidBurkett) [reply src/wallet/PegOut.cpp:29]: Fixed in the next PR.

**Pull request merged**